File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.py
import re

def solution(files):
    def split_file(file):
        original = file
        file = file.lower()
        num = re.search(r'[0-9]+', file).group()
        head, tail = re.split(num, file, maxsplit=1)
        return [head, int(num), original]

    fileInfo = [split_file(file) for file in files]
    
    fileInfo.sort(key = lambda x: (x[0], x[1]))

    return [file[2] for file in fileInfo]

File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.js
function solution(files) {
    const re = /^([a-zA-Z-\. ]+)([0-9]+)(.*)$/
    let dict = []
    files.forEach((v, i) => {
        let [file, head, num] = v.match(re)
        dict.push({file, head: head.toLowerCase(), num: parseInt(num), i})
    })

    return dict.sort((a, b) => {
        if (a.head > b.head) return  1
        if (a.head < b.head) return -1
        if (a.num > b.num) return  1
        if (a.num < b.num) return -1
        return a.i - b.i
    }).map(v => v.file)
}